File: solutions/c/grains/1/grains.c
#include "grains.h"

uint64_t square(uint8_t index) {

    if(index > 64 || index == 0) {
        return 0;
    } else if(index == 1) {
        return 1;
    }
    
    uint64_t prev = 1;
    uint64_t grainsInIndex = 0;
    
    for(uint8_t i = 2; i <= index; i++) {
        grainsInIndex = prev * 2; 
        prev = grainsInIndex;
    }

    return grainsInIndex;
}

uint64_t total(void) {

    uint64_t prev = 1;
    uint64_t totalGrains = 1;
    
    for(uint8_t i = 1; i < 64; i++) {

        prev *= 2;
        totalGrains += prev;
    }

    return totalGrains;
}

File: solutions/java/grains/1/src/main/java/Grains.java
import java.math.BigInteger;
import java.lang.Math;

class Grains {

    BigInteger grainsOnSquare(final int square) throws IllegalArgumentException {

        if(square > 64 || square <= 0) {
            
            throw new IllegalArgumentException("square must be between 1 and 64");
            
        } else if(square == 1) {
            
            return BigInteger.valueOf(1);
        }

        BigInteger prev = BigInteger.ONE;
        BigInteger grainsInIndex = BigInteger.ZERO;

        for(int i = 2; i <= square; i++) {
            
            grainsInIndex = prev.multiply(BigInteger.valueOf(2));
            
            prev = grainsInIndex;
        }
        
        return grainsInIndex;
    }

    BigInteger grainsOnBoard() {
        
        BigInteger prev = BigInteger.ONE;
        BigInteger totalGrains = BigInteger.ONE;
        

        for(int i = 1; i < 64; i++) {
            
            prev = prev.multiply(BigInteger.valueOf(2));
            totalGrains = totalGrains.add(prev);
        }

        return totalGrains;
    }

}
